torch/_inductor/codegen/cpp.py

```diff
@@ -1435,18 +1435,26 @@ def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
 
         if name not in V.graph.removed_buffers:
             var = self.args.output(name)
+            out_dtype = V.graph.get_dtype(name)
             if self.tiling_idx >= self.reduction_depth:
                 # Horizontal reduction
-                self.reduction_suffix.writeline(
-                    DeferredLine(name, f"{var}[{cexpr_index(index)}] = {tmpvar};")
-                )
-            else:
-                # Vertical reduction
                 self.reduction_suffix.writeline(
                     DeferredLine(
-                        name, f"{tmpvar_vec}.store({var} + {cexpr_index(index)});"
+                        name,
+                        f"{var}[{cexpr_index(index)}] = static_cast<{DTYPE_TO_CPP[out_dtype]}>({tmpvar});",
                     )
                 )
+            else:
+                # Vertical reduction
```

Comment:
BF16 legalization is to explicitly add `to_dtype` node to the fx graph because we expect all the dtype conversion should be represented in the fx graph for further optimization.